lib/ansible/plugins/inventory/foreman.py

```diff
@@ -121,134 +126,139 @@ def _get_session(self):
+            key = param['name']
+
+            # Try json
+            try:
+                value = json.loads(param['value'])
+            except ValueError:
+                # Not JSON, try YAML
+                pass
+            try:
+                value = yaml.load(param['value'])
+            except yaml.YAMLError:
+                # Not YAML or JSON, return plain string
+                key = value
 
-    def _get_hosts(self):
-        return self._get_json("%s/api/v2/hosts" % self.foreman_url)
+            self.inventory.set_variable(host['name'], key, value)
 
-    def _get_all_params_by_id(self, hid):
-        url = "%s/api/v2/hosts/%s" % (self.foreman_url, hid)
-        ret = self._get_json(url, [404])
-        if not ret or not isinstance(ret, MutableMapping) or not ret.get('all_parameters', False):
-            return {}
-        return ret.get('all_parameters')
+    def _get_facts_by_id(self, host):
 
-    def _get_facts_by_id(self, hid):
-        url = "%s/api/v2/hosts/%s/facts" % (self.foreman_url, hid)
-        return self._get_json(url)
+        s = self._get_session()
 
-    def _get_facts(self, host):
-        """Fetch all host facts of the host"""
+        ret = s.get("%s/api/v2/hosts/%s/facts" % (self.foreman_url, host))
+        ret.raise_for_status()
 
-        ret = self._get_facts_by_id(host['id'])
-        if len(ret.values()) == 0:
-            facts = {}
-        elif len(ret.values()) == 1:
-            facts = list(ret.values())[0]
-        else:
-            raise ValueError("More than one set of facts returned for '%s'" % host)
-        return facts
+        return ret.json()
 
-    def _populate(self):
+    def parse(self, inventory, loader, path, cache=True):
+
+        super(InventoryModule, self).parse(inventory, loader, path)
+
+        # read config from file, this sets 'options'
+        self._read_config_data(path)
+
+        # set config variables
+        self.foreman_url = self.get_option('url')
+        self.use_cache = cache and self.get_option('cache')
+        self.cache_key = self.get_cache_key(path)
 
-        for host in self._get_hosts():
+        hosts = self._get_all_hosts(self.foreman_url)
 
+        for host in hosts:
             if host.get('name'):
-                host_name = self.inventory.add_host(host['name'])
+                self.inventory.add_host(host['name'])
 
                 # create directly mapped groups
                 group_name = host.get('hostgroup_title', host.get('hostgroup_name'))
                 if group_name:
-                    group_name = to_safe_group_name('%s%s' % (self.get_option('group_prefix'), group_name.lower().replace(" ", "")))
-                    group_name = self.inventory.add_group(group_name)
-                    self.inventory.add_child(group_name, host_name)
+                    group_name = self.to_safe('%s%s' % (self.get_option('group_prefix'), group_name.lower()))
+                    self.inventory.add_group(group_name)
```

Comment:
add_group sanitizes names, which is why this was using the returned value before